### ISSUE TITLE
Reddit exotic oauth

### ISSUE THREAD

**Issue by nihaux** [FormidableLabs/react-native-app-auth#262]
Hello, im new to react-native and trying to implement reddit oauth thx to your library.
This might not be a feature request but more a question / feedback issue :)

After some experimentation I found that I the token request was failing with 401 response code.  

Actually took me some time to found that out because the exception catched in JS was "Error: Error not specified." which was not very helpful but it forced me to dig in the java code and learn how to debug the native module so yeah 👍 

So first feedback is, maybe more info could be added to this callback https://github.com/FormidableLabs/react-native-app-auth/blob/master/android/src/main/java/com/rnappauth/RNAppAuthModule.java#L230
=> in my case, errorDescription is null but the error property contains the http response code which is what I needed to understand the problem.  

So back to my problem, reddit seems to have an exotic oauth implementation where you need to pass a http basic auth header to the token request https://github.com/reddit-archive/reddit/wiki/oauth2#retrieving-the-access-token.  

From what I saw there is no way to pass additional headers to the token request, but maybe this could be a thing kinda like additionalParameters but additionalHeaders ? 
Or maybe there some other way that i did not think of ?

thx for reading and for the lib!

**Comment by kadikraman** [comment]
Hi @nihaux - I tried this on the weekend and I actually got a completely different error on Android 🙈- I think it may be due to the emulator I'm using. I have an idea, but I couldn't verify it, so here's the theory:

This would have to be a change in the native `AppAuth` libraries we use under the hood, but you can check that this _can_ work by editing the code for `AppAuth-Android`. [Here](https://github.com/openid/AppAuth-Android/commit/f29e3dfd7702e48ac4df145dfe76dfc976602918) is an example of when they added an `Accept: application/json` header to the request to support GitHub's exotic OAuth (you'd be surprised by all the big name companies that do not have spec compliant OAuth solutions).

Basically - open that file in Android Studio and add the http basic auth header the same way as in the example.

**Comment by nihaux** [comment]
Hello @kadikraman and thx for the answer :)  
Short version: you were right, it works by adding the header in AppAuth-Android lib 👍 
Long version:  
I was not able to  "open that file in Android Studio" as it was a compiled dependency and so read-only.
Instead, I followed this method (1) https://stackoverflow.com/a/48635089/632186  
And jump through some hoops (excluding dependency from react-native-app-auth, replacing unknown project variable in AppAuth-Android) to make it work but it the end it compiled :)   
Was there an easier way ?  
So what should I do next ? Make a PR to https://github.com/openid/AppAuth-Android ? I'm not a java dev but I could try something.

**Comment by kadikraman** [comment]
Awesome! I'm glad you could make it work 👏 - sorry I wasn't too sure on the specifics, since I hadn't tested it for myself.

They won't add the header just for this use-case (since it's outside the OAuth spec), but they may be open to adding a feature for custom headers, so you could open an issue to suggest that.

However, realistically I wouldn't expect AppAuth-Android to adapt to your use-case any time soon (AppAuth-iOS has had this in the backlog for many months), so for the time being, I'd just check in the custom linked library you have now (or you can fork the repo and host it in maven/gradle).

**Comment by nihaux** [comment]
Was about to open an issue on their repo when I stumbled upon this issue https://github.com/openid/AppAuth-Android/issues/174#issuecomment-282589679
which is the one that lead to the commit you referenced earlier.  

>  So, be wary that you may bump into other instances of where GitHub integration doesn't work well using AppAuth - we can tolerate making small changes to support them, but won't support completely off-spec behavior.

So I guess they wont accept to make custom headers a thing.  

But, the next comment is interesting
> Ah, also worth noting: you can actually already set extra headers and generally modify the requests sent in two ways:
> 1) Provide an AppAuthConfiguration with a custom ConnectionBuilder to AuthorizationService.
> 2) Provide a custom ClientAuthentication when performing the token request.

And after looking at the code of react-native-app-auth number 1) is actually the solution you use to allow for oauth to unsecure connections: [createConnectionBuilder](https://github.com/FormidableLabs/react-native-app-auth/blob/ec0d6e7bf3762805c21c942952d02126f39883e7/android/src/main/java/com/rnappauth/RNAppAuthModule.java#L405) [UnsafeConnectionBuilder](https://github.com/FormidableLabs/react-native-app-auth/blob/ec0d6e7bf3762805c21c942952d02126f39883e7/android/src/main/java/com/rnappauth/utils/UnsafeConnectionBuilder.java#L48)

So I guess my next question is would you be open to use the same technique to implement custom headers ?
Could be something like CustomConnectionBuilder, which proxy ::openConnection to the real ConnectionBuilder but adds custom headers on the way ?

**Comment by kadikraman** [comment]
Oh yes, you're right, I forgot about that!

I'd be happy to merge an "android-only" option for adding custom headers, if you'd be up for doing it.

An iOS solution may or may not follow. At the moment, GitHub auth won't work on iOS, because they didn't want to add an accept header as the Android library did, stating that it is "out of scope" (which is fair) - but they did say they'd be open for a solution that adds custom headers some time in the future. For background, here the issue that sparked the discussion: https://github.com/FormidableLabs/react-native-app-auth/issues/194

**Comment by nihaux** [comment]
Hello, so I made a [PR](https://github.com/FormidableLabs/react-native-app-auth/pull/263) for this.  
Besides the code in itself, I made 2 changes which are not directly related to the issue, tell me if I should revert them before squashing.
In order to be able to build the android part of the project, I had to put this repositories in the build.gradle (otherwise had some 404 on some dependencies).
Same thing for the gradle version in gradle-properties, had to upgrade it otherwise the build process throwed an error "you should upgrade etc". 
Other than that I tried it in my project and it's working properly :)

**Comment by nihaux** [comment]
Sooo I read the oauth spec, and it turns out that the way reddit does thing is actually spec compliant ^^.
AND there is actually a way to do send basic auth headers during token request in [App-Auth-Android](https://github.com/openid/AppAuth-Android#utilizing-client-secrets-dangerous).
AND it's implemented in [react-native-app-auth](https://github.com/FormidableLabs/react-native-app-auth/blob/master/android/src/main/java/com/rnappauth/RNAppAuthModule.java#L235)

Only difference, and the reason why it does not work in my case, is that reddit wants you to send the Basic header [even when you don't have a client secret](https://github.com/reddit-archive/reddit/wiki/OAuth2#retrieving-the-access-token).
Maybe if I pass an empty string for clientSecret it will do the trick. I'll try this.

So the last order of business is the validity of my PR, assuming the "empty string as clientSecret" trick work, I will not need to pass custom headers. But maybe someone else will and have a use for it ?

**Comment by imns** [comment]
For anyone coming here trying to get Reddit Oauth2 to work on IOS, as @nihaux said you have to pass an empty string for `clientSecret`.  [AppAuth-iOS](https://github.com/openid/AppAuth-iOS/blob/65ef95c6b38c7889fdda7a22f242fa0156fd43b2/Source/OIDTokenRequest.m#L277) Checks for a client secret, but allows an empty string.

**Comment by kadikraman** [comment]
Thanks for the tip @imns - that works! I'll add it to the Readme.
